src/Raven.Server/Documents/Indexes/MapReduce/AutoMapReduceIndex.cs

```diff
@@ -75,24 +60,29 @@ protected override IIndexingWork[] CreateIndexWorkExecutors()
         
         public override IDisposable InitializeIndexingWork(TransactionOperationContext indexContext)
         {
-            _mapReduceWorkContext.MapEntriesTable = GetMapEntriesTable(indexContext.Transaction.InnerTransaction);
+            _mapReduceWorkContext.MapEntries = GetMapEntriesTree(indexContext.Transaction.InnerTransaction);
 
             return _mapReduceWorkContext;
         }
 
         public override unsafe void HandleDelete(DocumentTombstone tombstone, IndexWriteOperation writer, TransactionOperationContext indexContext, IndexingStatsScope stats)
         {
-            var mapEntry = GetMapEntryForDocument(tombstone.Key);
+            var documentMapEntries = _mapReduceWorkContext.MapEntries.FixedTreeFor(tombstone.Key, sizeof(ulong));
 
-            if (mapEntry == null)
+            if (documentMapEntries.NumberOfEntries == 0)
                 return;
 
-            var state = GetReduceKeyState(mapEntry.ReduceKeyHash, indexContext, create: false);
+            foreach (var mapEntry in GetMapEntriesForDocument(tombstone.Key, documentMapEntries))
+            {
+                var state = GetReduceKeyState(mapEntry.ReduceKeyHash, indexContext, create: false);
+                
+                fixed (long* ptr = &mapEntry.Id)
+                    state.Tree.Delete(new Slice((byte*)ptr, sizeof(long)));
 
-            var storageId = mapEntry.StorageId;
+                _mapReduceWorkContext.EntryDeleted(mapEntry);
+            }
 
-            _mapReduceWorkContext.MapEntriesTable.Delete(storageId);
-            state.Tree.Delete(new Slice((byte*)&storageId, sizeof(long))); 
+            _mapReduceWorkContext.MapEntries.DeleteFixedTreeFor(tombstone.Key, sizeof(ulong));
         }
 
         public override unsafe void HandleMap(Document document, IndexWriteOperation writer, TransactionOperationContext indexContext, IndexingStatsScope collectionScope)
```

Comment:
You are still only processing the mapped result one at a time.
You need to assume that the index can return multiple entries from the map